exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/SqlPatternContainsMatcher.java

```diff
@@ -19,44 +19,283 @@
+      final byte secondPattByte = patternArray[1];
 
       // simplePattern string has meta characters i.e % and _ and escape characters removed.
       // so, we can just directly compare.
-      for (int patternIndex = 0; patternIndex < patternLength; patternIndex++) {
-        if (patternByteBuffer.get(patternIndex) != drillBuf.getByte(start + txtIndex + patternIndex)) {
-          continue outer;
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        final byte firstInByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte != firstInByte) {
+          continue;
+        } else {
+          final byte secondInByte = drillBuf.getByte(start + idx +1);
+
+          if (secondInByte == secondPattByte) {
+            return 1;
+          }
         }
       }
+      return 0;
+    }
+  }
+
+  /** Handles patterns with length three */
+  private final class Matcher3 extends MatcherFcn {
 
-      return 1;
+    private Matcher3() {
+      super();
     }
 
-    return  0;
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+      final int lengthToProcess  = end - start -2;
+      final byte firstPattByte   = patternArray[0];
+      final byte secondPattByte  = patternArray[1];
+      final byte thirdPattByte   = patternArray[2];
+
+      // simplePattern string has meta characters i.e % and _ and escape characters removed.
+      // so, we can just directly compare.
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        final byte inputByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte != inputByte) {
+          continue;
+        } else {
+          final byte secondInByte = drillBuf.getByte(start + idx +1);
+          final byte thirdInByte  = drillBuf.getByte(start + idx +2);
+
+          if (secondInByte == secondPattByte && thirdInByte == thirdPattByte) {
+            return 1;
+          }
+        }
+      }
+      return 0;
+    }
+  }
+
+  /** Handles patterns with arbitrary length */
+  private final class MatcherN extends MatcherFcn {
+
+    private MatcherN() {
+      super();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+
+      if (patternLength == 0) {
+        return 1;
+      }
+
+      final int lengthToProcess = end - start - patternLength + 1;
+      int patternIndex          = 0;
+      final byte firstPattByte  = patternArray[0];
+
+      // simplePattern string has meta characters i.e % and _ and escape characters removed.
+      // so, we can just directly compare.
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        final byte inputByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte == inputByte) {
+          for (patternIndex = 1; patternIndex < patternLength; ++patternIndex) {
+            final byte currInByte   = drillBuf.getByte(start + idx + patternIndex);
```

Comment:
you might end up accessing drillBuf out of bound if idx+patternIndex > length of drillBuf.

Reply:
This will not happen because of the outer-loop condition; idx < lengthToProcess which guarantees that we always have enough bytes to fully match the pattern; example:
- input-len = 6; pattern = 4 this means lengthToProcess = 3 (6-4+1)
- index 0, 1, and 2 are allowed since we have at least 4 bytes we can match (1 within the outer-loop and 3 in the inner one); index 3 will be rejected since we have only three bytes left (b3, b4, and b5)
- If no match, then